reporter/src/main/kotlin/reporters/NoticeByPackageReporter.kt

```diff
@@ -101,7 +102,7 @@ class NoticeByPackageProcessor(input: ReporterInput) : AbstractNoticeReporter.No
             .filter { (license, _) ->
                 input.licenseTextProvider.hasLicenseText(license).also {
                     if (!it) {
-                        NoticeByPackageProcessor.log.warn {
+                        NoticeByPackageProcessor.log.warnOnce {
```

Comment:
Would it make sense to implement this instead by introducing a small code block (maybe at the beginning of the reporter) which collects the license IDs which do not have a text and logs all license IDs in a single place?

That'd be simpler to understand imo.

Reply:
Maybe the implementation at *this* place would be simpler to understand then, but this is just an example usage to showcase the feature; I actually plan to make of the log-once functions in more places throughout the code base. Hence the generic (but slightly harder to understand) implementation.

Reply:
Actually I was not worried much about the complexity of the code, but about the complexity of the behavior and the logs. The logging becomes stateful and thus it's more tricky to read the logs because the log reader does not always know the state the logging function was in. 

Further it seems to introduce the hidden assumption that only one notice by package report is created throughout the lifetime of the whole process. That is, for all subsequently created notices in the same process the log output would be missing.

What do you think?

Reply:
That's a good point, at least associating the store with a specific instance could prevent some unwanted side effects. This could be done with a `WeakHashMap` to reduce memory usage. But I also wonder if it isn't a big performance hit, if these functions are used more often and every log statement has to be compared with a potentially long list of former statements.

Reply:
> Further it seems to introduce the hidden assumption that only one notice by package report is created throughout the lifetime of the whole process (or is it per class instance?).

De-duplication of log messages indeed happens per class *instance*.

> But I also wonder if it isn't a big performance hit, if these functions are used more often and every log statement has to be compared with a potentially long list of former statements.

I agree that this new feature must be used sparsely, to avoid both performance and memory penalties. But from a user (person who reads the logs) this is a feature I very much like to have.

Reply:
> De-duplication of log messages indeed happens per class _instance_.

How that? The extension property `KotlinLogger.statements` is static, so the same map is used for all instances of `KotlinLogger`.

Reply:
Hmm, then extension properties work different from what I thought. I'll take a look again.

Reply:
You would have to "simulate" the missing backing field with a map similar to `loggerOfClass`.

Reply:
Sure, I can do that, but I'm still surprised how extension properties apparently are implemented. I knew the *internal implementation* uses static functions, but I thought this function would get `this` passed as an argument and dispatch to the respective instance. Somewhat disappointing that this is not the case.

Reply:
Ah, no, I simply misremembered how I've implemented `loggerOfClass` :laughing: I'll come up with a fix.